[RFC] Enable -Wshorten-64-to-32

azat opened the pull request:
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

This is to avoid issues like #42146 

### This patchset adds `-Wshorten-64-to-32` to the build arguments, and also fixes all places that has such conversions.

_Note, in theory this can break something in a backward incompatible way, even though I tried to do this carefully._

### Here are some steps that was used to avoid hidden issues:
- I've added a test to compare aggregate states of functions against `22.9` - 
 [`test_functions.py::test_aggregate_states`](https://github.com/azat/ClickHouse/blob/1fd4483bff4dc6573245f61209f6350d0bff8c3b/tests/integration/test_backward_compatibility/test_functions.py#L31)
- One more test to compare result of string functions against `22.9` -  [`test_functions.py::test_string_functions`](https://github.com/azat/ClickHouse/blob/1fd4483bff4dc6573245f61209f6350d0bff8c3b/tests/integration/test_backward_compatibility/test_functions.py#L121)
- Encryption functions uses `safe_cast` over `static_cast` (that throws if boundaries exceeds)

### About conversion:
- there was some places that uses `getUInt` over `getUInt64` for `Poco::AbstractConfiguration`, but they were not critical, since using there value > `UINT_MAX` was unlikely.
- `Poco`/`select`/`epoll` (non `epoll_pwait2()`), uses `int` for timeout, however ClickHouse does not have such type for settings

**And I would love to have extra eyes here.**

azat committed:
Enable -Wshorten-64-to-32

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Fixes for -Wshorten-64-to-32

- lots of static_cast
- add safe_cast
- types adjustments
  - config
  - IStorage::read/watch
  - ...
- some TODO's (to convert types in future)

P.S. That was quite a journey...

v2: fixes after rebase
v3: fix conflicts after #42308 merged
Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented:
So up to the fork point of this topic branch everything compiled successfully on linux, however there can be new changes. Let's see.

azat committed:
Fix -Wshorten-64-to-32 for musl

On musl uint_fast32_t is uint32_t, while for glibc it is unsigned long
long.

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Fix -Wshorten-64-to-32 with USE_BLAKE3

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Fix -Wshorten-64-to-32 in self extracting executable

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Fix -Wshorten-64-to-32 for darwin builds

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Fix -Wshorten-64-to-32 for amd64sse2 build

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
tests/gtest_DecimalFunctions: fix operator<<

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
tests/gtest_DecimalFunctions: fix due to incorrect cast in wide_integer

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Add a note about incorrect wide_integer with Decimal

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Fix -Wshorten-64-to-32 in examples

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
tests/gtest_compressionCodec: fix for darwin

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented on src/Common/HashTable/Hash.h:
```diff
@@ -48,33 +48,36 @@ inline DB::UInt64 intHash64(DB::UInt64 x)
+/// NOTE: Intel intrinsic can be confusing.
+/// - https://code.google.com/archive/p/sse-intrinsics/wikis/PmovIntrinsicBug.wiki
+/// - https://stackoverflow.com/questions/15752770/mm-crc32-u64-poorly-defined
+inline DB::UInt32 intHashCRC32(DB::UInt64 x)
```
Looks unsafe, and I guess it is better to keep signature as before and fix the callers. Thoughts?

rschu1ze replied:
Yes, that looks unsafe, please keep the signature.

azat replied:
Actually there was not too much with previous signature, will be fixed shortly (need to fix conflicts with upstream, sigh).

azat committed:
tests/gtest_compressionCodec: Fix UBSAN report about signed integer overflow

UBSAN report:

    $ UBSAN_SYMBOLIZER_PATH=/usr/bin/llvm-symbolizer-14 UBSAN_OPTIONS=print_stacktrace=1 ./unit_tests_dbms --gtest_filter=*Gorilla*
    ../src/Compression/tests/gtest_compressionCodec.cpp:1216:47: runtime error: signed integer overflow: 23 * 100000000 cannot be represented in type 'int'
        #0 0x14f67fd1 in auto (anonymous namespace)::$_6::operator()(int) const::'lambda'(auto)::operator()<int>(auto) const build_docker/../src/Compression/tests/gtest_compressionCodec.cpp:1216:47
        #1 0x14f67fd1 in (anonymous namespace)::CodecTestSequence (anonymous namespace)::generateSeq<long, (anonymous namespace)::$_6::operator()(int) const::'lambda'(auto), int, int>((anonymous namespace)::$_6::operator()(int) const::'lambda'(auto), char const*, int, int) build_docker/../src/Compression/tests/gtest_compressionCodec.cpp:394:36
        #2 0x14f67fd1 in auto (anonymous namespace)::GCompatibilityTestSequence<long>() build_docker/../src/Compression/tests/gtest_compressionCodec.cpp:1224:12
        #3 0x14f3c7f3 in (anonymous namespace)::gtest_GorillaCodecTestCompatibility_EvalGenerator_() build_docker/../src/Compression/tests/gtest_compressionCodec.cpp:1227:1
        #4 0x14f6bdb5 in testing::internal::ParameterizedTestSuiteInfo<(anonymous namespace)::CodecTestCompatibility>::RegisterTests() build_docker/../contrib/googletest/googletest/include/gtest/internal/gtest-param-util.h:553:45
        #5 0x27d87988 in testing::internal::ParameterizedTestSuiteRegistry::RegisterTests() build_docker/../contrib/googletest/googletest/include/gtest/internal/gtest-param-util.h:726:24
        #6 0x27d87988 in testing::internal::UnitTestImpl::RegisterParameterizedTests() build_docker/../contrib/googletest/googletest/src/gtest.cc:2805:34
        #7 0x27d87988 in testing::internal::UnitTestImpl::PostFlagParsingInit() build_docker/../contrib/googletest/googletest/src/gtest.cc:5492:5
        #8 0x27d9d002 in void testing::internal::InitGoogleTestImpl<char>(int*, char**) build_docker/../contrib/googletest/googletest/src/gtest.cc:6499:22
        #9 0x14fd5495 in main build_docker/../src/Coordination/tests/gtest_coordination.cpp:2189:5
        #10 0x7f8c29005209  (/lib/x86_64-linux-gnu/libc.so.6+0x29209) (BuildId: 71a7c7b97bc0b3e349a3d8640252655552082bf5)
        #11 0x7f8c290052bb in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x292bb) (BuildId: 71a7c7b97bc0b3e349a3d8640252655552082bf5)
        #12 0x14ce356d in _start (/work1/azat/tmp/42190/unit_tests_dbms+0x14ce356d) (BuildId: 482550e3f8d45f06e8c7f8147f427ee798c1f645)

    SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior ../src/Compression/tests/gtest_compressionCodec.cpp:1216:47 in

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
tests/gtest_compressionCodec: fix UBSAN report to avoid test failure

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Fix clang-tidy performance-inefficient-vector-operation

By some reason it appears only after static_cast<> was added [1]:

    /build/src/Processors/Formats/Impl/AvroRowInputFormat.cpp
    Oct 18 01:03:56 /build/src/Processors/Formats/Impl/AvroRowInputFormat.cpp:351:21: error: 'push_back' is called inside a loop; consider pre-allocating the container capacity before the loop [performance-inefficient-vector-operation,-warnings-as-errors]
    Oct 18 01:03:56                     symbols.push_back(root_node->nameAt(i));
    Oct 18 01:03:56                     ^
    Oct 18 01:03:56 /build/src/Processors/Formats/Impl/AvroRowInputFormat.cpp:511:17: error: 'push_back' is called inside a loop; consider pre-allocating the container capacity before the loop [performance-inefficient-vector-operation,-warnings-as-errors]
    Oct 18 01:03:56                 union_skip_fns.push_back(createSkipFn(root_node->leafAt(i)));
    Oct 18 01:03:56                 ^
    Oct 18 01:03:56 /build/src/Processors/Formats/Impl/AvroRowInputFormat.cpp:552:17: error: 'push_back' is called inside a loop; consider pre-allocating the container capacity before the loop [performance-inefficient-vector-operation,-warnings-as-errors]
    Oct 18 01:03:56                 field_skip_fns.push_back(createSkipFn(root_node->leafAt(i)));
    Oct 18 01:03:56                 ^
    Oct 18 01:03:56 197965 warnings generated.

  [1]: https://s3.amazonaws.com/clickhouse-builds/42190/453d91fa3539882dcef1d5ecd5097747499572d8/clickhouse_special_build_check/report.html

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented:
There is one minor conflict that pops up because one of patches had been applied individually via #42217. But in general this is ready, except for one major comment. It will be great if someone can take a look too.

azat committed:
Add backward compatiblity check for some functions

* State of aggregate functions

This test goes though all aggregate functions that:
- has only one argument
- support string as an argument

And do a simple check by creating the aggregate state with one string.

Yes this is not covers everything (does not cover functions with
different number of arguments, types, different states in case of
multiple values - uniqCombined, but as for uniqCombined it will be
checked via uniqHLL12), but at least something.

And about type, String had been selected, since it more likely that
there will be used some hash function internally.

* Function(String)

Check every function which has only one argument, and it's type is
String.

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

alexey-milovidov commented on base/base/itoa.h:
```diff
@@ -122,7 +122,7 @@ QuotientAndRemainder<N> static inline split(UnsignedOfSize<N> value)
 
     UnsignedOfSize<N> quotient = (division.multiplier * (UnsignedOfSize<2 * N>(value) + division.add)) >> division.shift;
-    UnsignedOfSize<N / 2> remainder = value - quotient * pow10<UnsignedOfSize<N / 2>>(N);
+    UnsignedOfSize<N / 2> remainder = static_cast<UnsignedOfSize<2 * N>>(value - quotient * pow10<UnsignedOfSize<N / 2>>(N));
```
Why here is 2 * N instead of N / 2 as in the result?

azat replied:
That was an error, nice catch! Rebased.

azat committed:
Fix type error in itoa (wrong static_cast<>)

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented:
>Conflicting files
>src/Functions/DateTimeTransforms.h

Rebased.

rschu1ze commented on src/IO/readDecimalText.h:
```diff
@@ -106,7 +106,7 @@ inline bool readDigits(ReadBuffer & buf, T & x, uint32_t & digits, int32_t & exp
 
                     // TODO: accurate shift10 for big integers
-                    x *= intExp10OfSize<T>(places);
+                    x *= intExp10OfSize<typename T::NativeType>(places);
```
`x` is a `T`, not a `T::NativeType`.

azat replied:
Same as above.

rschu1ze commented on src/Interpreters/ClusterProxy/SelectStreamFactory.cpp:
```diff
@@ -131,7 +131,7 @@ void SelectStreamFactory::createForShard(
         }
 
-        UInt32 local_delay = replicated_storage->getAbsoluteDelay();
+        UInt64 local_delay = replicated_storage->getAbsoluteDelay();
```
`getAbsoluteDelay()` returns `time_t`.

azat replied:
This is to compare it with `settings.max_replica_delay_for_distributed_queries`

rschu1ze commented on src/Interpreters/convertFieldToType.cpp:
```diff
@@ -88,7 +88,7 @@ Field convertIntToDecimalType(const Field & from, const DataTypeDecimal<T> & typ
         throw Exception("Number is too big to place in " + type.getName(), ErrorCodes::ARGUMENT_OUT_OF_BOUND);
 
-    T scaled_value = type.getScaleMultiplier() * static_cast<T>(value);
+    T scaled_value = type.getScaleMultiplier() * T(static_cast<typename T::NativeType>(value));
```
I don't really understand the logic here.

azat replied:
This is for Decimal, static_cast<Decimal<Int>> still produce warning

rschu1ze commented on src/Core/MySQL/MySQLReplication.cpp:
```diff
@@ -625,7 +625,7 @@ namespace MySQLReplication
                                     UInt32 val = 0;
                                     readBigEndianStrict(payload, reinterpret_cast<char *>(&val), 4);
-                                    res *= intExp10OfSize<DecimalType>(digits_per_integer);
+                                    res *= intExp10OfSize<typename DecimalType::NativeType>(static_cast<int>(digits_per_integer));
```
What was wrong with `intExp10OfSize<DecimalType>`? `res` is a `DecimalType`.

rschu1ze commented on src/Core/MySQL/MySQLReplication.cpp:
```diff
@@ -638,7 +638,7 @@ namespace MySQLReplication
                                     UInt32 val = 0;
                                     readBigEndianStrict(payload, reinterpret_cast<char *>(&val), 4);
-                                    res *= intExp10OfSize<DecimalType>(digits_per_integer);
+                                    res *= intExp10OfSize<typename DecimalType::NativeType>(static_cast<int>(digits_per_integer));
```
same

rschu1ze commented on src/Core/MySQL/MySQLReplication.cpp:
```diff
@@ -651,7 +651,7 @@ namespace MySQLReplication
                                     {
                                         readBigEndianStrict(payload, reinterpret_cast<char *>(&val), to_read);
-                                        res *= intExp10OfSize<DecimalType>(compressed_decimals);
+                                        res *= intExp10OfSize<typename DecimalType::NativeType>(static_cast<int>(compressed_decimals));
```
same

rschu1ze commented on src/Common/HyperLogLogCounter.h:
```diff
@@ -264,7 +264,8 @@ enum class HyperLogLogMode
 template <
     UInt8 precision,
-    typename Hash = IntHash32<UInt64>,
+    typename Key = UInt64,
```
I dont' get the logic behind the changes in this file.

azat replied:
There is no way to get type of `Key` for hashing, I've tried to do some meta programming, but failed because operator() cannot be static. So that's why separate template argument had been introduced instead, to cast using this type.

rschu1ze replied:
Alright, thanks.

rschu1ze commented on src/IO/FileEncryptionCommon.cpp:
```diff
@@ -84,10 +85,13 @@ namespace
             out.nextIfAtEnd();
+
             size_t part_size = std::min(size - in_size, out.available());
+            part_size = std::min<size_t>(part_size, INT_MAX);
```
Should that not throw an exception instead of silently truncate? (--> safe_cast)

azat replied:
This function has a loop over input, so should safe.

rschu1ze replied:
Hm, okay.

azat committed:
Review fixes

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented on programs/odbc-bridge/ColumnInfoHandler.cpp:
```diff
@@ -183,7 +183,10 @@ void ODBCColumnsInfoHandler::handleRequest(HTTPServerRequest & request, HTTPServ
+        WriteBufferFromHTTPServerResponse out(
+            response,
+            request.getMethod() == Poco::Net::HTTPRequest::HTTP_HEAD,
+            static_cast<unsigned>(keep_alive_timeout));
```
Actually right now cast is required, I forgot that I've changed WriteBufferFromHTTPServerResponse

azat commented on src/Backups/BackupSettings.cpp:
```diff
@@ -72,6 +71,7 @@ namespace
     M(String, host_id) \
     M(String, coordination_zk_path) \
     M(OptionalUUID, backup_uuid)
+    /// M(Int64, compression_level)
```
This is deeply in the defines. It can be fixed by checking types or cast for any setting or by using Int64 in BackupSettings.h but all of this looks worse. Thoughts?

rschu1ze replied:
Another option is to turn the warning off around this code using a pragma.

azat replied:
I thought about this, but it is too tempting to have such example in the code :)

rschu1ze replied:
Fine with me then.

azat commented on src/Client/ClientBase.cpp:
```diff
@@ -1866,7 +1868,9 @@ bool ClientBase::executeMultiQuery(const String & all_queries_text)
-                    adjustQueryEnd(this_query_end, all_queries_end, global_context->getSettingsRef().max_parser_depth);
+                    adjustQueryEnd(
+                        this_query_end, all_queries_end,
+                        static_cast<unsigned>(global_context->getSettingsRef().max_parser_depth));
```
Better to change the interface of adjustQueryEnd, since mac_par

azat commented on src/Columns/ColumnArray.cpp:
```diff
@@ -569,8 +569,8 @@ void ColumnArray::expand(const IColumn::Filter & mask, bool inverted)
-    int index = mask.size() - 1;
-    int from = offsets_data.size() - 1;
+    ssize_t index = mask.size() - 1;
+    ssize_t from = offsets_data.size() - 1;
```
There is a check for negative value in the loop, prefer to make less changes.

azat commented on src/Columns/ColumnFixedString.cpp:
```diff
@@ -277,8 +277,8 @@ void ColumnFixedString::expand(const IColumn::Filter & mask, bool inverted)
 
-    int index = mask.size() - 1;
-    int from = size() - 1;
+    ssize_t index = mask.size() - 1;
```
Same as for arrays

azat commented on src/Columns/ColumnFixedString.cpp:
```diff
@@ -277,8 +277,8 @@ void ColumnFixedString::expand(const IColumn::Filter & mask, bool inverted)
-    int index = mask.size() - 1;
-    int from = size() - 1;
+    ssize_t index = mask.size() - 1;
+    ssize_t from = size() - 1;
```
Same as for arrays

azat commented on src/Columns/ColumnString.cpp:
```diff
@@ -168,8 +168,8 @@ void ColumnString::expand(const IColumn::Filter & mask, bool inverted)
 
-    int index = mask.size() - 1;
-    int from = offsets_data.size() - 1;
+    ssize_t index = mask.size() - 1;
```
Same as for arrays

azat commented on src/Columns/ColumnString.cpp:
```diff
@@ -168,8 +168,8 @@ void ColumnString::expand(const IColumn::Filter & mask, bool inverted)
-    int index = mask.size() - 1;
-    int from = offsets_data.size() - 1;
+    ssize_t index = mask.size() - 1;
+    ssize_t from = offsets_data.size() - 1;
```
Same as for arrays

azat commented on src/Storages/System/StorageSystemStackTrace.cpp:
```diff
@@ -324,7 +324,7 @@ Pipe StorageSystemStackTrace::read(
 
             sig_value.sival_int = sequence_num.load(std::memory_order_acquire);
-            if (0 != ::sigqueue(tid, sig, sig_value))
+            if (0 != ::sigqueue(static_cast<int>(tid), sig, sig_value))
```
Well since there are other platforms I prefer to keep it `int` instead of `pid_t` to avoid extra failures. Objections?
P.S. at least `pid_t` not `__pid_t`

rschu1ze replied:
Yepp, it is okay to keep `int'.

azat commented on src/Functions/CustomWeekTransforms.h:
```diff
@@ -134,14 +134,17 @@ struct WeekTransformer
         vec_to.resize(size);
 
         for (size_t i = 0; i < size; ++i)
+        {
```
Usually I'm using the same style as in the file, even though I prefer to use parentheses always.
However, this for loop has if/else, and even though it is work as expected, it is even more easier to make an error here:

So that's why I prefer:

```c
for (;;)
{
  if (foo)
    make_foo();
  else
    make_bar();
}
```

Over:

```c
for (;;)
  if (foo)
    make_foo();
  else
    make_bar();
```

Note, that I did not put extra parentheses for if/else.

But anyway, let me know what you thing, I'm to revert this change.

P.S. and if I did not have to change the lines inside if/else I wouldn't touch the whole hunk.

rschu1ze replied:
It's okay and I see your argument.

azat commented on src/Common/parseRemoteDescription.cpp:
```diff
@@ -82,8 +82,8 @@ std::vector<String> parseRemoteDescription(const String & description, size_t l,
         {
-            int cnt = 1;
-            int last_dot = -1; /// The rightmost pair of points, remember the index of the right of the two
+            ssize_t cnt = 1;
```
Yes it can. I prefer to use ssize_t/size_t for standard (read: non db specific) code like this one, but if you thing Int64 should be used here, fine by me, let me know.

rschu1ze replied:
`ssize_t` is okay for now. I was maybe sceptical because I see this type used quite rarely in ClickHouse.

azat commented on src/Common/parseRemoteDescription.cpp:
```diff
@@ -82,8 +82,8 @@ std::vector<String> parseRemoteDescription(const String & description, size_t l,
-            int cnt = 1;
-            int last_dot = -1; /// The rightmost pair of points, remember the index of the right of the two
+            ssize_t cnt = 1;
+            ssize_t last_dot = -1; /// The rightmost pair of points, remember the index of the right of the two
```
Same as above. Will resolve this and change both if needed.

azat commented on src/Formats/ProtobufSerializer.cpp:
```diff
@@ -1029,7 +1029,7 @@ namespace
                 {
                     this->write_function = [this](NumberType value) { writeInt(enumDataTypeValueToProtobufEnumValue(value)); };
-                    this->read_function = [this]() -> NumberType { return protobufEnumValueToEnumDataTypeValue(readInt()); };
+                    this->read_function = [this]() -> NumberType { return protobufEnumValueToEnumDataTypeValue(static_cast<NumberType>(readInt())); };
```
I used `NumberType` since function returns it, while readInt returns Int64

azat commented on src/Server/TCPHandler.h:
```diff
@@ -159,7 +159,7 @@ class TCPHandler : public Poco::Net::TCPServerConnection
     UInt64 client_version_minor = 0;
     UInt64 client_version_patch = 0;
-    UInt64 client_tcp_protocol_version = 0;
+    UInt32 client_tcp_protocol_version = 0;
```
The protocol part located in ClientInfo.h - https://github.com/ClickHouse/ClickHouse/blob/440af00c74dc66b1f2479e9895b1f0cfd99978bc/src/Interpreters/ClientInfo.h#L81

azat commented on src/Storages/MergeTree/ReplicatedMergeTreeLogEntry.cpp:
```diff
@@ -240,7 +240,7 @@ void ReplicatedMergeTreeLogEntryData::readText(ReadBuffer & in)
                 if (checkString("merge_type: ", in))
                 {
-                    UInt64 value;
+                    UInt32 value;
```
Merge type is very small, and checkAndGetMergeType() (used just below) accept UInt32, the whole MergeType is enum of 3 elements actually, so I wouldn't expect to have something different there, except in case of corruption, in this case checkAndGetMergeType will throw an exception.

rschu1ze replied:
You are right, I misread the code.

azat commented on src/Interpreters/ThreadStatusExt.cpp:
```diff
@@ -55,12 +55,12 @@ void ThreadStatus::applyQuerySettings()
         LOG_TRACE(log, "Setting nice to {}", new_os_thread_priority);
 
-        if (0 != setpriority(PRIO_PROCESS, thread_id, new_os_thread_priority))
+        if (0 != setpriority(PRIO_PROCESS, static_cast<unsigned>(thread_id), new_os_thread_priority))
```
Same as with pid, to avoid issues on other platforms. Objections?

rschu1ze replied:
Nope, `unsigned` is fine. After all, future readers can see our discussion here to find out why no `id_t` is used :smile:

azat commented on src/Interpreters/ThreadStatusExt.cpp:
```diff
@@ -349,7 +349,7 @@ void ThreadStatus::detachQuery(bool exit_if_already_detached, bool thread_exits)
         LOG_TRACE(log, "Resetting nice");
 
-        if (0 != setpriority(PRIO_PROCESS, thread_id, 0))
+        if (0 != setpriority(PRIO_PROCESS, static_cast<int>(thread_id), 0))
```
Same, will resolve, and change both places if needed.

azat commented on src/Coordination/SummingStateMachine.cpp:
```diff
@@ -71,10 +71,10 @@ void SummingStateMachine::createSnapshotInternal(nuraft::snapshot & s)
 
     // Maintain last 3 snapshots only.
-    int num = snapshots.size();
+    ssize_t num = snapshots.size();
```
Does ClickHouse non-posix compatible? There are other places with this type, if you thing it is better to use Int64 let me know.

rschu1ze replied:
CH only runs on posix systems, so we are generally good. `ssize_t` here is fine, usage in non-system code is however discouraged (I would say).

azat commented on src/Storages/HDFS/ReadBufferFromHDFS.cpp:
```diff
@@ -90,7 +91,7 @@ struct ReadBufferFromHDFS::ReadBufferFromHDFSImpl : public BufferWithOwnMemory<S
         }
 
-        int bytes_read = hdfsRead(fs.get(), fin, internal_buffer.begin(), num_bytes_to_read);
+        int bytes_read = hdfsRead(fs.get(), fin, internal_buffer.begin(), safe_cast<int>(num_bytes_to_read));
```
You are indeed right, it is very unlikely to get here > INT_MAX, however it can be a huge problem if here incorrect number of bytes will be read, hence `safe_cast`.

azat commented on src/Storages/Hive/StorageHive.cpp:
```diff
@@ -809,7 +809,7 @@ Pipe StorageHive::read(
 
     if (num_streams > sources_info->hive_files.size())
-        num_streams = sources_info->hive_files.size();
+        num_streams = static_cast<unsigned>(sources_info->hive_files.size());
```
I've changed `IStorage::read` to accept `size_t` for `num_streams`, hence this cast is obsolete.

azat commented on src/Common/RadixSort.h:
```diff
@@ -273,13 +273,13 @@ struct RadixSort
         {
             /// Replace the histograms with the accumulated sums: the value in position i is the sum of the previous positions minus one.
-            size_t sums[NUM_PASSES] = {0};
+            CountType sums[NUM_PASSES] = {0};
```
Later it uses values to index `histograms`, which is of `CountType` type.

rschu1ze replied:
Ah, okay.

azat commented on src/Common/ThreadPool.cpp:
```diff
@@ -87,7 +87,7 @@ void ThreadPoolImpl<Thread>::setQueueSize(size_t value)
 template <typename Thread>
 template <typename ReturnType>
-ReturnType ThreadPoolImpl<Thread>::scheduleImpl(Job job, int priority, std::optional<uint64_t> wait_microseconds, bool propagate_opentelemetry_tracing_context)
+ReturnType ThreadPoolImpl<Thread>::scheduleImpl(Job job, ssize_t priority, std::optional<uint64_t> wait_microseconds, bool propagate_opentelemetry_tracing_context)
```
Reading you comments about this on and on, and have more to say:

```
azat:.../worktrees/clickhouse-upstream (build/shorten-64-to-32) [1]$ gg -c '\<ssize_t\>' src | wc -l
117
azat:.../worktrees/clickhouse-upstream (build/shorten-64-to-32) [1]$ gg -c '\<Int64\>' src | wc -l
356
```

So maybe you right, and better to use `Int64`, but maybe better to convert it all at once separatelly.

rschu1ze replied:
Coming back to earlier discussion, it is not that ssize_t is wrong, but most people will probably be more familiar with the C++ stdlib types. In any case, converting existing ssize_t to Int64 should definitely not be done in this PR :-)

azat commented on src/Functions/FunctionsRound.h:
```diff
@@ -436,7 +436,7 @@ class DecimalRoundingImpl
         if (scale_arg > 0)
         {
-            auto scale = intExp10OfSize<T>(scale_arg);
+            auto scale = intExp10OfSize<NativeType>(scale_arg);
```
This is for `Decimal`, `static_cast<Decimal<Int>>` still produce warning

azat committed:
Revert changes of hash functions signatures and fix callers

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Merge remote-tracking branch 'upstream/master' into build/shorten-64-to-32

Conflicts:
- src/Common/intExp.h
- src/Functions/DateTimeTransforms.h
- src/Functions/FunctionsConversion.h

azat commented:
@rschu1ze thank you, great review!

I've addressed most of the comments, other has some followups. Please take a look!

>src/Common/intExp.h
>src/Functions/DateTimeTransforms.h
>src/Functions/FunctionsConversion.h

Rebased.

rschu1ze reviewed (approved):
I'd say this PR is good to go.

It will run into frequent merge conflicts though. If you like to turn tables around and let everyone else have merge conflicts ^^, I suggest to
- revert the removal of `no_warning(shorten-64-to-32)` in warnings.cmake to remove the risk of compile errors in other ongoing PRs. Then merge this PR as soon as possible.
- eventually do a small PR where `no_warning(shorten-64-to-32)` is deleted in a couple of days.

But that's up to you.

azat commented:
Usually there is less activity on weekends, so I will try to address further issues ASAP on weekends, if it will not be ready by weekends, then maybe I will go simpler way.

azat committed:
Merge remote-tracking branch 'upstream/master' into build/shorten-64-to-32

Conflicts:
- src/Functions/GregorianDate.h

azat commented:
>Conflicting files
>src/Functions/GregorianDate.h

Fixed, conflicts with #42464 (sigh).

azat committed:
Fix one clang-tidy issue

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented:
>[ClickHouse special build check — 8/9 artifact groups are OK](https://s3.amazonaws.com/clickhouse-builds/42190/dbb4f8216f413dc2ee96449b51199ed915880f3c/clickhouse_special_build_check/report.html)

```
/build/src/Core/iostream_debug_helpers.h -Werror -nostdinc++ -std=gnu++20 -MD -MT src/CMakeFiles/dbms.dir/DataTypes/getLeastSupertype.cpp.o -MF src/CMakeFiles/dbms.dir/DataTypes/getLeastSupertype.cpp.o.d -o src/CMakeFiles/dbms.dir/DataTypes/getLeastSupertype.cpp.o -c /build/src/DataTypes/getLeastSupertype.cpp
Oct 22 07:21:45 /build/src/DataTypes/getLeastSupertype.cpp:541:13: error: use range-based for loop instead [modernize-loop-convert,-warnings-as-errors]
Oct 22 07:21:45             for (size_t i = 0; i < int_ids.size(); ++i)
Oct 22 07:21:45             ^   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Oct 22 07:21:45                 (auto & int_id : int_ids)
Oct 22 07:21:45 178717 warnings generated.
```

Fixed.

azat committed:
Merge remote-tracking branch 'upstream/master' into build/shorten-64-to-32

Conflicts:
- src/Interpreters/ProcessList.cpp

azat commented:
>Conflicting files
>src/Interpreters/ProcessList.cpp

And one more time...

P.S. apparently my hope that there will be less changes in upstream on weekends failed.

rschu1ze commented:
@azat Do you have merge rights? If no, pls ping me - I can also monitor CI results and merge when everything is green.

azat commented:
> Do you have merge rights?

No (I thought only people from ClickHouse Inc have them).

> pls ping me - I can also monitor CI results and merge when everything is green.

Will do, thanks for the assistance!

azat commented:
_(I will update this comment to include details about other failures)_

>[Stateless tests (asan) [1/2] — fail: 1, passed: 2247, skipped: 16](https://s3.amazonaws.com/clickhouse-test-reports/42190/56bc85746fa0b553e43c2253250404cfcca46855/stateless_tests__asan__[1/2].html)

- `02458_datediff_date32` - is flaky - https://github.com/ClickHouse/ClickHouse/pull/42308/files#r1002536937

azat commented on src/Core/tests/gtest_DecimalFunctions.cpp:
```diff
@@ -144,6 +146,17 @@ TEST_P(DecimalUtilsSplitAndCombineForDateTime64Test, getFractionalPartDateTime64
 
 }
 
+namespace std
```
There is one report from `clang-tidy`

```
Oct 22 20:03:51 /build/src/Core/tests/gtest_DecimalFunctions.cpp:152:16: error: modification of 'std' namespace can result in undefined behavior [cert-dcl58-cpp,-warnings-as-errors]
```

Which should be fixed by the following patch:

```suggestion
namespace std // NOLINT(cert-dcl58-cpp)
```

azat commented:
So apart this `clang-tidy` error everything looks OK. And suggestion should address the `clang-tidy` error.

azat commented:
@rschu1ze what do you think about applying suggestion and merge?

azat committed:
Merge remote-tracking branch 'upstream/master' into build/shorten-64-to-32

azat committed:
Fix clang-tidy issue in gtest_DecimalFunctions

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Fix clang-tidy cert-dcl58-cpp in gtest_DecimalFunctions

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Merge remote-tracking branch 'upstream/master' into build/shorten-64-to-32

* upstream/master:
  Fix strange code in date monotonicity
  Add a test
  Fix error
  Add tests
  Remove trash

azat commented:
>[Stress test (tsan) — Sanitizer assert (in stderr.log)](https://s3.amazonaws.com/clickhouse-test-reports/42190/bcc8d9f9d9570fdfab141d5830bef044e1676749/stress_test__tsan_.html)

- same - #42583

Felixoid commented:
> P.S. Now it retries `clang-tidy` again, though there is no sense in this, @Felixoid you may be interesting in this, also note, that `clang-tidy` does not benefit from `ccache` and it takes ~4 hours in total, so it is better to fix this to reduce resource usage.

I've tried fixing it in https://github.com/ClickHouse/ClickHouse/pull/39652 and https://github.com/ClickHouse/ClickHouse/pull/39709, but was failed so far. The benefit of improving it is obvious, but https://github.com/ClickHouse/clickhouse-repos-manager/tree/working-system is the highest priority for me now.

azat commented:
>I've tried fixing it in https://github.com/ClickHouse/ClickHouse/pull/39652 and https://github.com/ClickHouse/ClickHouse/pull/39709, but was failed so far.

`ccache` simply does not support `clang-tidy`, so those PRs does not makes any difference.

Felixoid commented:
Interesting. It looks there is a solution still https://stackoverflow.com/a/62934377/3706827

azat commented:
This is not about making ccache cache the clang-tidy results, since it is not possible, ccache supports only compiling one unit (`-c`) however clang-tidy is not a compiler but a linting tool.

```
[2022-10-24T12:10:38.159371 15216] Command line: ccache clang-tidy ../test.c
[2022-10-24T12:10:38.159432 15216] Working directory: /tmp/ccache-clang-tidy/.cmake
[2022-10-24T12:10:38.159449 15216] Compiler type: clang
[2022-10-24T12:10:38.159532 15216] No -c option found
[2022-10-24T12:10:38.159553 15216] Failed; falling back to running the real compiler
[2022-10-24T12:10:38.159573 15216] Executing /usr/bin/clang-tidy ../test.c
```

There is an external tool, written in go, https://github.com/ejfitzgerald/clang-tidy-cache, however it is not maintained anymore.

nvartolomei commented:
I'm a bit confused about the test added here and why it passes.

I downloaded the debug build from build checks and run the query:

```
dev :) select value from system.build_options where name = 'GIT_HASH'

SELECT value
FROM system.build_options
WHERE name = 'GIT_HASH'

Query id: ce0cb34c-97cd-4f60-8d8e-f620b688fbc3

┌─value────────────────────────────────────┐
│ 22b90f9ceaf359c9dbf4f6df30eab391bc92dfb1 │
└──────────────────────────────────────────┘

1 row in set. Elapsed: 0.012 sec.

dev :) SELECT hex(initializeAggregation('minState', 'foo'))

SELECT hex(initializeAggregation('minState', 'foo'))

Query id: 93c9d437-0969-4001-886f-f88091c9a3db

┌─hex(initializeAggregation('minState', 'foo'))─┐
│ 03000000666F6F                                │
└───────────────────────────────────────────────┘
```

Then I did the same with 22.9 docker image and I get different results 🤔 

```
c26e531fed5b :) select version()

SELECT version()

Query id: b75ff50e-6ebb-4ba5-9fb1-485ee939b455

┌─version()─┐
│ 22.9.6.20 │
└───────────┘

1 row in set. Elapsed: 0.001 sec.

c26e531fed5b :) SELECT value
                FROM system.build_options
                WHERE name = 'GIT_HASH'

SELECT value
FROM system.build_options
WHERE name = 'GIT_HASH'

Query id: 20ab9384-c137-458d-bf2a-f43913668cc4

┌─value────────────────────────────────────┐
│ ef6343f957987e23133abee5a89a7114dde7a780 │
└──────────────────────────────────────────┘

1 row in set. Elapsed: 0.001 sec.

c26e531fed5b :) SELECT hex(initializeAggregation('minState', 'foo'))

SELECT hex(initializeAggregation('minState', 'foo'))

Query id: b5b999e3-ab00-4804-ac9e-00cb1b2734b4

┌─hex(initializeAggregation('minState', 'foo'))─┐
│ 04000000666F6F00                              │
└───────────────────────────────────────────────┘

1 row in set. Elapsed: 0.001 sec.
```

How did this test pass here?

```
03000000666F6F != 04000000666F6F00
```

nvartolomei commented:
I think I see how...

The 22.6 image was broken at the time as well. Correct?

22.9.1 borked. 22.9.6 fixed.

azat commented:
@nvartolomei looks like here you are referencing the regression that had been introduced by #41431, and you are wondering why this patchset, that includes that PR, does not fail.

The test had been passing because it uses image with a bug for the older clickhouse (22.9).

>The 22.6 image was broken at the time as well. Correct?

Not exactly, the switch to 22.6 had been done in #43038

nvartolomei commented:
@azat thanks for details and for links! 🙇
